In [1]:
import pandas as pd
import numpy as np

# Raw - Mobility & Air quality

Importación de datos de calidad de aire y mobilidad

inicialmente se leen los datos crudos desde los archivos descargables de las  respectivas fuentes, se procede a compilar todos los datos de calidad de aire en 3 dataframe diferentes (Train, Test y Validation) que se convierten posteriormente a un archivo HDF5 el cual permite recolectar la información de manera más eficiente facilitando la lectura para proximas ocasiones, de igual maneraa se procede con los datos de mobilidad 

In [2]:
"""
directory_path = 'C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train'
dfs = []

for filename in os.listdir(directory_path):
    if filename.endswith('.csv'):
        file_path = os.path.join(directory_path, filename)
        df = pd.read_csv(file_path)
        dfs.append(df)

Train_air = pd.concat(dfs, ignore_index=True)

# Save the combined DataFrame to an HDF5 file
Train_air.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df', mode='w') 
"""

"\ndirectory_path = 'C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train'\ndfs = []\n\nfor filename in os.listdir(directory_path):\n    if filename.endswith('.csv'):\n        file_path = os.path.join(directory_path, filename)\n        df = pd.read_csv(file_path)\n        dfs.append(df)\n\nTrain_air = pd.concat(dfs, ignore_index=True)\n\n# Save the combined DataFrame to an HDF5 file\nTrain_air.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df', mode='w') \n"

In [3]:
# Load the DataFrame from the HDF5 file
Train_air = pd.read_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/train_air_dataframe.h5', key='df')

Los datos de train van desde 2021-01-01 hasta 2023-05-01

In [4]:
Train_air.head()

Fecha_Hora  codigoSerial  pm25  calidad_pm25  pm10  calidad_pm10  \
0  2021-01-01 00:00:00            12  31.0           1.0  49.0           1.0   
1  2021-01-01 01:00:00            12  41.0           1.0  51.0           1.0   
2  2021-01-01 02:00:00            12  38.0           1.0  51.0           1.0   
3  2021-01-01 03:00:00            12  30.0           1.0  42.0           1.0   
4  2021-01-01 04:00:00            12  29.0           1.0  39.0           1.0   

      pm1  calidad_pm1      no  calidad_no  ...   p_ssr  calidad_p_ssr  \
0 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
1 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
2 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
3 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   
4 -9999.0          1.0 -9999.0        4.31  ... -9999.0            1.0   

   pliquida_ssr  calidad_pliquida_ssr  rglobal_ssr  calidad_rglobal_ssr  \
0       -9999.0                   1.0      -9999.0                  1.0   
1       -9999.0                   1.0      -9999.0                  1.0   
2       -9999.0                   1.0      -9999.0                  1.0   
3       -9999.0                   1.0      -9999.0                  1.0   
4       -9999.0                   1.0      -9999.0                  1.0   

   taire10_ssr  calidad_taire10_ssr  vviento_ssr  calidad_vviento_ssr  
0      -9999.0                  1.0      -9999.0                  1.0  
1      -9999.0                  1.0      -9999.0                  1.0  
2      -9999.0                  1.0      -9999.0                  1.0  
3      -9999.0                  1.0      -9999.0                  1.0  
4      -9999.0                  1.0      -9999.0                  1.0  

[5 rows x 36 columns]

In [5]:
Train_air.columns

Index(['Fecha_Hora', 'codigoSerial', 'pm25', 'calidad_pm25', 'pm10',
       'calidad_pm10', 'pm1', 'calidad_pm1', 'no', 'calidad_no', 'no2',
       'calidad_no2', 'nox', 'calidad_nox', 'ozono', 'calidad_ozono', 'co',
       'calidad_co', 'so2', 'calidad_so2', 'pst', 'calidad_pst', 'dviento_ssr',
       'calidad_dviento_ssr', 'haire10_ssr', 'calidad_haire10_ssr', 'p_ssr',
       'calidad_p_ssr', 'pliquida_ssr', 'calidad_pliquida_ssr', 'rglobal_ssr',
       'calidad_rglobal_ssr', 'taire10_ssr', 'calidad_taire10_ssr',
       'vviento_ssr', 'calidad_vviento_ssr'],
      dtype='object')

In [6]:
Air = Train_air.copy()
Air = Train_air[['Fecha_Hora', 'codigoSerial', 'pm25', 'calidad_pm25']]

Del dataset anterior únicamente se tomaran en cuenta las columnas de fecha_hora, codigoSerial, pm25 y calidad_pm25, ya que, las demás columnas al tener valores de -9999 y calidad 1 significan que los sensores están fuera de servicio y por la descripción del dataset se evidencia que solo las columnas de pm25 tienen un comportamiento diferente al mencionado

In [7]:
#Ponemos bajo un mismo formato a la columnda de Fecha_Hora
Air.loc[:,'Fecha_Hora'] = pd.to_datetime(Air['Fecha_Hora'], format='%Y-%m-%d %H:%M:%S')

In [8]:
# Establecer la columna 'fecha' como índice
Air.set_index('Fecha_Hora', inplace=True)

In [9]:
Air.head(10)

codigoSerial  pm25  calidad_pm25
Fecha_Hora                                           
2021-01-01 00:00:00            12  31.0           1.0
2021-01-01 01:00:00            12  41.0           1.0
2021-01-01 02:00:00            12  38.0           1.0
2021-01-01 03:00:00            12  30.0           1.0
2021-01-01 04:00:00            12  29.0           1.0
2021-01-01 05:00:00            12  33.0           1.0
2021-01-01 06:00:00            12  26.0           1.0
2021-01-01 07:00:00            12  30.0           1.0
2021-01-01 08:00:00            12  31.0           1.0
2021-01-01 09:00:00            12  24.0           1.0

La columna codigoSerial representa el código de la estación con esta podemos saber donde se encuentra ubicada, para poder comparar los datos con el dataset de mobilidad

Serial 12
Comuna 10 
Estación Tráfico Centro - La Candelaria

Serial 79
Comuna 16
Belén Altavista - I.E Pedro Octavio Amado
Belén - I.E Pedro Justo Barrio

Serial 80
Comuna 8
Villahermosa - Planta de Producción Agua Potable EPM

Serial 83   
Comuna 16
Belén - I.E Pedro Justo Berrio

Serial 84
Comuna 14
El Poblado - I.E INEM Sede Santa Catalina

Serial 85
Comuna 10
CEN-TRAF - San Cristobal - Parque Biblioteca Fernando Botero

Serial 86
Comuna 4
Aranjuez - I.E Ciro Mendia

https://www.medellin.gov.co/es/conoce-algunos-datos-generales-de-la-ciudad/


Nos fijamos en la cantidad de datos que hay para cada estación, notamos que la estación 86 tiene varios faltantes a comparación de las demás estaciones

In [10]:
Air['codigoSerial'].value_counts()

codigoSerial
12    21144
79    21144
80    21144
83    21144
84    21144
85    21144
86    21144
Name: count, dtype: int64

Creamos una función para calcular el porcentaje de datos nulos en el dataset

In [11]:
def missing_values(df):
    Column_Names = df.columns

    Total_rows = df.shape[0]

    missing_values = df.isnull().sum()

    Percent_missing = (missing_values / Total_rows)*100
    
    result = pd.DataFrame({
        'Column_Name': Column_Names,
        'Missing_Values': missing_values.values,
        'Total_Rows': Total_rows,
        'Percent_Missing': Percent_missing.values
    })

    return result

Cambiamos la distribución de la tabla con ayuda de una tabla pivote, así por cada una de las estaciones tenemos una columna para calidad y otra para la magnitud de pm2.5

In [12]:
Air_organized = Air.pivot_table(index=Air.index, columns='codigoSerial', values=['pm25', 'calidad_pm25'], aggfunc = 'mean')

In [13]:
Air_organized.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 21144 entries, 2021-01-01 00:00:00 to 2023-05-31 23:00:00
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   (calidad_pm25, 12)  21144 non-null  float64
 1   (calidad_pm25, 79)  21144 non-null  float64
 2   (calidad_pm25, 80)  21144 non-null  float64
 3   (calidad_pm25, 83)  21144 non-null  float64
 4   (calidad_pm25, 84)  21144 non-null  float64
 5   (calidad_pm25, 85)  21144 non-null  float64
 6   (calidad_pm25, 86)  21144 non-null  float64
 7   (pm25, 12)          21144 non-null  float64
 8   (pm25, 79)          21144 non-null  float64
 9   (pm25, 80)          21144 non-null  float64
 10  (pm25, 83)          21144 non-null  float64
 11  (pm25, 84)          21144 non-null  float64
 12  (pm25, 85)          21144 non-null  float64
 13  (pm25, 86)          21144 non-null  float64
dtypes: float64(14)
memory usage: 2.4 MB


In [14]:
missing_values(Air_organized)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0   (calidad_pm25, 12)               0       21144              0.0
1   (calidad_pm25, 79)               0       21144              0.0
2   (calidad_pm25, 80)               0       21144              0.0
3   (calidad_pm25, 83)               0       21144              0.0
4   (calidad_pm25, 84)               0       21144              0.0
5   (calidad_pm25, 85)               0       21144              0.0
6   (calidad_pm25, 86)               0       21144              0.0
7           (pm25, 12)               0       21144              0.0
8           (pm25, 79)               0       21144              0.0
9           (pm25, 80)               0       21144              0.0
10          (pm25, 83)               0       21144              0.0
11          (pm25, 84)               0       21144              0.0
12          (pm25, 85)               0       21144              0.0
13          (pm25, 86)               0       21144              0.0

Resulta que este dataset tiene la particularidad que los datos nulos o erroneos no son NaN, sino que se conocen por el valor de la calidad del datos, si es de1 el valor es real y ha sido revisado, si es un numero diferente, puede ser dudoso, malo, el sensor estar fuera de servicio, por lo que unicamente tomaremos los datos con valores de calidad de 2.6

In [15]:
# Revisamos la cantidad de datos con diferentes calificaciones de calidad
# Seleccionar todas las columnas con el nombre 'calidad_pm25' en el primer nivel del MultiIndex
calidad_pm25_columns = Air_organized.xs('calidad_pm25', level=0, axis=1)

# Iterar sobre cada columna (por código serial)
for column in calidad_pm25_columns.columns:
    # Contar la frecuencia de cada valor único
    value_counts = calidad_pm25_columns[column].value_counts()
    
    # Mostrar los resultados
    print(f"Columna (código serial): {column}")
    print(value_counts)
    print("\n")


Columna (código serial): 12
12
1.0000    19267
4.3100      473
2.6000      373
2.1000      322
2.3000      233
3.1000      233
4.2985      115
2.2000       88
3.0000       32
2.4000        4
2.0000        3
4.2995        1
Name: count, dtype: int64


Columna (código serial): 79
79
1.00    19952
2.30      252
3.10      238
3.30      186
4.32      120
4.31      115
3.00      109
2.20       99
2.40       70
2.00        1
3.20        1
2.60        1
Name: count, dtype: int64


Columna (código serial): 80
80
1.00    19745
3.00      429
2.30      243
2.40      177
3.10      171
4.31      117
2.20      115
3.30       78
4.32       68
3.20        1
Name: count, dtype: int64


Columna (código serial): 83
83
1.00    20113
2.30      256
4.31      197
3.10      160
4.32      107
2.20      102
3.30       87
2.40       66
3.00       52
3.20        4
Name: count, dtype: int64


Columna (código serial): 84
84
1.00    19967
2.30      204
4.31      179
3.00      175
3.10      172
2.40      144
3.30     

In [16]:
Air_organized.describe()

calidad_pm25                                            \
codigoSerial            12            79            80            83   
count         21144.000000  21144.000000  21144.000000  21144.000000   
mean              1.183014      1.117000      1.128331      1.104229   
std               0.643953      0.508954      0.509751      0.496410   
min               1.000000      1.000000      1.000000      1.000000   
25%               1.000000      1.000000      1.000000      1.000000   
50%               1.000000      1.000000      1.000000      1.000000   
75%               1.000000      1.000000      1.000000      1.000000   
max               4.310000      4.320000      4.320000      4.320000   

                                                                pm25  \
codigoSerial            84            85            86            12   
count         21144.000000  21144.000000  21144.000000  21144.000000   
mean              1.115980      1.167397      1.110002   -193.593431   
std               0.509823      0.641769      0.507426   1488.366251   
min               1.000000      1.000000      1.000000  -9999.000000   
25%               1.000000      1.000000      1.000000     18.000000   
50%               1.000000      1.000000      1.000000     24.000000   
75%               1.000000      1.000000      1.000000     32.000000   
max               4.320000      4.320000      4.320000    995.000000   

                                                                         \
codigoSerial            79             80             83             84   
count         21144.000000   21144.000000   21144.000000   21144.000000   
mean            547.413988     295.496136     439.403983     366.326301   
std            7629.090483    5813.275188    7217.585234    6587.224924   
min           -9999.000000   -9999.000000   -9999.000000   -9999.000000   
25%              12.961550      10.697525      10.923350       9.590198   
50%              18.384550      15.996800      18.155250      14.647850   
75%              24.938900      22.470475      27.872825      20.360800   
max           99999.000000  104533.000000  117237.000000  117429.000000   

                                            
codigoSerial             85             86  
count          21144.000000   21144.000000  
mean             987.977519     397.001109  
std            10665.482414    6940.904490  
min            -9999.000000   -9999.000000  
25%                9.839388      12.116600  
50%               14.716250      17.618100  
75%               20.667425      23.770575  
max           100117.000000  108778.000000

Desde el diccionario de datos del SIATA menciona que valores de calidad mayores a 2.6 no son confiables por lo que se eliminan de mi dataset en aras de mejorar la calidad de mis datos y por ende la predicción

In [17]:
prueba = Air_organized.copy()

# Filtrar todas las columnas de 'calidad_pm25'
calidad_columns = prueba.xs('calidad_pm25', level=0, axis=1)

# Iterar sobre cada código serial en las columnas 'calidad_pm25'
for codigoSerial in calidad_columns:
    # Reemplazar los valores en 'pm25' si 'calidad_pm25' es menor a 2.6
    prueba.loc[prueba[('calidad_pm25', codigoSerial)] > 1, ('pm25', codigoSerial)] = np.nan
    #prueba.loc[prueba[('calidad_pm25', codigoSerial)] > 1, ('calidad_pm25', codigoSerial)] = np.nan


In [18]:
missing_values(prueba)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0   (calidad_pm25, 12)               0       21144         0.000000
1   (calidad_pm25, 79)               0       21144         0.000000
2   (calidad_pm25, 80)               0       21144         0.000000
3   (calidad_pm25, 83)               0       21144         0.000000
4   (calidad_pm25, 84)               0       21144         0.000000
5   (calidad_pm25, 85)               0       21144         0.000000
6   (calidad_pm25, 86)               0       21144         0.000000
7           (pm25, 12)            1877       21144         8.877223
8           (pm25, 79)            1192       21144         5.637533
9           (pm25, 80)            1399       21144         6.616534
10          (pm25, 83)            1031       21144         4.876088
11          (pm25, 84)            1177       21144         5.566591
12          (pm25, 85)            1522       21144         7.198260
13          (pm25, 86)            1084       21144         5.126750

In [19]:
# Crear columnas auxiliares de año, mes, día a partir del índice
prueba['Año'] = prueba.index.year
prueba['Mes'] = prueba.index.month
prueba['Día'] = prueba.index.day

# Calcular la media horaria para cada día específico
# Esto incluye agrupar por año, mes, día  y luego calcular la media
medias_horarias = prueba.groupby(['Año', 'Mes', 'Día']).transform('mean')

# Rellenar valores faltantes usando las medias horarias calculadas
prueba.fillna(medias_horarias, inplace=True)

# Eliminar las columnas auxiliares si no son necesarias
prueba.drop(['Año', 'Mes', 'Día'], axis=1, inplace=True)

In [20]:
missing_values(prueba)

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0   (calidad_pm25, 12)               0       21144         0.000000
1   (calidad_pm25, 79)               0       21144         0.000000
2   (calidad_pm25, 80)               0       21144         0.000000
3   (calidad_pm25, 83)               0       21144         0.000000
4   (calidad_pm25, 84)               0       21144         0.000000
5   (calidad_pm25, 85)               0       21144         0.000000
6   (calidad_pm25, 86)               0       21144         0.000000
7           (pm25, 12)             144       21144         0.681044
8           (pm25, 79)             144       21144         0.681044
9           (pm25, 80)              96       21144         0.454030
10          (pm25, 83)             144       21144         0.681044
11          (pm25, 84)              96       21144         0.454030
12          (pm25, 85)             216       21144         1.021566
13          (pm25, 86)              96       21144         0.454030

Al ya no necesitar los datos de calidad_pm25 pues hemos ya sacado/borrado las magnitudes asociadas a aquellas calidades diferentes de 1, sólo sacamos la descripción de las magnitudes por cada estación

In [21]:
prueba['pm25'].describe()

codigoSerial            12            79            80            83  \
count         21000.000000  21000.000000  21048.000000  21000.000000   
mean             25.908674     19.457681     17.193898     20.190187   
std              10.270955      8.918790      8.621304     11.504186   
min               2.000000      0.000000      0.000000      0.000000   
25%              19.000000     13.341625     11.259875     11.204906   
50%              25.000000     18.488200     16.549350     18.293750   
75%              32.000000     24.552950     21.792325     27.334246   
max              92.000000     75.933400    112.304000     78.207900   

codigoSerial            84            85            86  
count         21048.000000  20928.000000  21048.000000  
mean             15.606293     15.746424     18.770021  
std               8.070167      7.698368      8.714379  
min               0.000000      0.000000      0.000000  
25%               9.852787     10.160050     12.832100  
50%              14.700000     14.767300     17.677150  
75%              20.096025     19.937575     23.486575  
max              75.675600     65.921400    110.534000

In [22]:
missing_values(prueba['pm25'])

Column_Name  Missing_Values  Total_Rows  Percent_Missing
0          12             144       21144         0.681044
1          79             144       21144         0.681044
2          80              96       21144         0.454030
3          83             144       21144         0.681044
4          84              96       21144         0.454030
5          85             216       21144         1.021566
6          86              96       21144         0.454030

Para los datos nulos, se tienen 2 alternativas

1. **Forward or backward fill:** Use the value from the previous or next observation to fill null values. This method is especially useful for time series data.

df.fillna(method='ffill', inplace=True, axis=0)

2. **Interpolation:** Estimate missing values based on other values in the dataset. Pandas provides various interpolation methods such as linear, polynomial, or spline interpolation.

df.interpolate(method='linear', inplace=True, axis=0)

Inicialmente se llevara a cabo **Forward or backward fill** pues es más acorde al tipo de datos

In [23]:
# Forward fill null values
intento = prueba['pm25'].copy()
intento.ffill(axis=0, inplace=True)


In [24]:
intento.describe()

codigoSerial            12            79            80            83  \
count         21144.000000  21144.000000  21144.000000  21144.000000   
mean             25.905891     19.498485     17.183040     20.210572   
std              10.242473      8.946704      8.603723     11.475611   
min               2.000000      0.000000      0.000000      0.000000   
25%              19.000000     13.375250     11.533498     11.276625   
50%              25.000000     18.515100     16.515104     18.335950   
75%              32.000000     24.570050     21.768900     27.352025   
max              92.000000     75.933400    112.304000     78.207900   

codigoSerial            84            85            86  
count         21144.000000  21144.000000  21144.000000  
mean             15.614804     15.733463     18.766878  
std               8.053626      7.661464      8.694778  
min               0.000000      0.000000      0.000000  
25%               9.880947     10.216000     12.843075  
50%              14.727350     14.738050     17.677150  
75%              20.051225     19.877825     23.473657  
max              75.675600     65.921400    110.534000

| Rango de PM2.5 (µg/m³) | Nivel de Salud                     |
|--------------------------|------------------------------------|
| 0 - 50                   | Saludable                          |
| 51 - 100                 | Moderado                           |
| 101 - 150                | Poco saludable para grupos sensibles |
| 151 - 200                | Poco saludable para la población en general |
| Más de 200               | Muy poco saludable o peligroso     |


In [25]:
intento.to_hdf('C:/Users/MARIA PAULA/Downloads/EAFIT/Proyecto de Grado/Air Quality/Train/Train_set_1A.h5', key='df', mode='w', format='table')

# Load the DataFrame from the HDF5 file
#Train_Air = pd.read_hdf('Train_set_1A.h5', key='df')